internal/tracker/downstream.go

```diff
@@ -0,0 +1,51 @@
+package tracker
+
+import (
+	"sync"
+)
+
+// DownstreamConns tracks connections per downstream based on a
+// unique string identifier.
+// DownstreamConns safe for concurrent use.
+type DownstreamConns struct {
+	// mu protects the resources of DownstreamConns
+	mu sync.Mutex
+
+	// connCounts is a map of downstream to a count of connections
+	connCounts map[string]uint32
+}
+
+// NewDownstreamConns initializes and returns a DownstreamConns with
+func NewDownstreamConns() *DownstreamConns {
+	return &DownstreamConns{
+		connCounts: map[string]uint32{},
+	}
+}
+
+// TryBeginConnection checks if a downstream is below a maximum
+// and if so records an additional connection for the client.
+// If the client has no history, a new count will be started.
+// The return indicates if the new connection should be allowed.
+func (t *DownstreamConns) TryBeginConnection(downstream string, max uint32) bool {
```

Comment:
Why the max is passed as an argument? It sounds like the maximum connection count should be a part of the configuration. With this design, you need to keep a separate map with the maximum allowed connection count and pass it every time you call this method.

Reply:
Yes. I saw this as a simplification of tracker.downstreamConn. I also will need that map you mentioned to exist already for configuration to the library, so I plan to use that as a lookup. It also loosely enables future opportunities for changing rate limits without restarting the library/server, a feature I have commonly used in ratelimiters.